apps/framework-cli/src/infrastructure/redis/redis_client.rs

```diff
@@ -61,768 +176,788 @@ impl Default for RedisConfig {
+/// - Consider connection pooling for high-throughput scenarios
+/// - The mock implementation uses RwLock for better read concurrency
+/// - For further optimization, consider implementing a custom connection
+///   wrapper with
+///   specialized read-only operations that don't require exclusive access
+///
+/// ## Usage Example
+///
+/// ```rust
+/// let config = RedisConfig::default();
+/// let client = RedisClient::new("my-service".to_string(), config).await?;
+///
+/// // Register a message handler
+/// client.register_message_handler(Arc::new(|msg| {
+///     println!("Received message: {}", msg);
+/// })).await;
+///
+/// // Start background tasks
+/// client.start_periodic_tasks();
+///
+/// // Use the client for various Redis operations
+/// client.set_with_service_prefix("my-key", "my-value").await?;
+/// ```
+///
+/// ## Error Handling Strategy
+///
+/// This client follows a graceful degradation approach to Redis connectivity issues:
+///
+/// - **Critical operations**: Methods that affect application correctness (like lock management)
+///   propagate errors to allow callers to handle them appropriately.
+///
+/// - **Non-critical operations**: Methods like messaging and presence updates fail silently
+///   with warnings logged, preventing Redis issues from cascading into application failures.
+///
+/// - **Fallback mode**: When Redis is completely unavailable, a mock implementation
+///   provides partial functionality for essential operations.
+///
+/// ## Reconnection Strategy
+///
+/// The client automatically monitors the Redis connection health and attempts reconnection:
+///
+/// - Health checks run every 5 seconds
+/// - Reconnection attempts use exponential backoff
+/// - Connection state is tracked atomically and can be queried via `is_connected()`
+/// - During reconnection attempts, operations either use the fallback or fail gracefully
 pub struct RedisClient {
-    connection: Arc<Mutex<ConnectionManager>>,
-    pub_sub: Arc<Mutex<ConnectionManager>>,
-    redis_config: RedisConfig,
-    service_name: String,
-    instance_id: String,
-    locks: HashMap<String, RedisLock>,
-    presence_task: Option<JoinHandle<()>>,
-    message_callbacks: Arc<Mutex<Vec<MessageCallback>>>,
-    listener_task: Mutex<Option<JoinHandle<()>>>,
-    connection_state: Arc<AtomicBool>,
+    pub config: RedisConfig,
+    pub service_name: String,
+    pub instance_id: String,
+    pub connection_manager: ConnectionManagerWrapper,
+    pub leadership_manager: LeadershipManager,
+    pub presence_manager: PresenceManager,
+    pub messaging_manager: MessagingManager,
+
+    pub message_callbacks: Arc<RwLock<Vec<MessageCallback>>>,
+    pub listener_task: Arc<RwLock<Option<JoinHandle<()>>>>,
+    pub presence_task: Arc<RwLock<Option<JoinHandle<()>>>>,
+    pub connection_monitor_task: Arc<RwLock<Option<JoinHandle<()>>>>,
+    pub fallback: Option<MockRedisClient>,
+
+    pub locks: Arc<RwLock<std::collections::HashMap<String, (String, i64)>>>,
```

Comment:
That's probably a lefover in the rollback, we don't need that right? since the locks are on the redis side

Reply:
Redis does manage the locks on the server side, having this local cache does allow for local checks without making network calls to Redis for every lock status query.  The issue is keeping those in sync - which the current code doesn't do well and may be more trouble than it's worth.  So, I'm refactoring this out in the next commit.